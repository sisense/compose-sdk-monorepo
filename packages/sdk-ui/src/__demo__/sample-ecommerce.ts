import { measureFactory } from '@sisense/sdk-data';

import { Commerce, Brand, Country, Category, DataSource } from './sample-ecommerce-autogenerated';

export { Commerce, Brand, Category, Country, DataSource };

/** adding predefined set of measures */
export const Measures = {
  SumCost: measureFactory.sum(Commerce.Cost).format('0,0$'),
  AvgCost: measureFactory.average(Commerce.Cost).format('0,0$'),
  SumRevenue: measureFactory.sum(Commerce.Revenue, 'Total Revenue').format('0,0$'),
  Quantity: measureFactory.sum(Commerce.Quantity, 'Total Quantity').format('0,0'),
  CountBrands: measureFactory.countDistinct(Commerce.BrandID).format('0,0'),
};
