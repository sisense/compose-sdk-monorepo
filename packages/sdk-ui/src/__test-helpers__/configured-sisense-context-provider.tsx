import { mockToken, mockUrl } from '@/__mocks__/msw';
import { SisenseContextProviderProps } from '@/props';
import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';

const contextProviderProps: SisenseContextProviderProps = {
  url: mockUrl,
  token: mockToken,
  defaultDataSource: 'Sample ECommerce',
  appConfig: {
    queryCacheConfig: { enabled: false },
    trackingConfig: { enabled: false },
  },
};

export const ConfiguredSisenseContextProvider: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  return <SisenseContextProvider {...contextProviderProps}>{children}</SisenseContextProvider>;
};
