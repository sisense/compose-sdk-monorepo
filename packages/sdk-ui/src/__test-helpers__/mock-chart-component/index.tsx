// Mocks the Chart component and provides a way to access the rendered charts.
// Use this to simulate user interactions with the chart while actual Chart component is mocked.

import { ChartProps } from '@/props';
import { RenderedChartModel } from './rendered-chart-model';

export type ChartMocksManager = {
  /**
   * Replace the actual Chart component with a mocked version.
   */
  mockChartComponent: () => void;
  /**
   * Restore the actual Chart component.
   */
  unmockChartComponent: () => void;
  /**
   * List of models of rendered charts.
   * Use these models to simulate user interactions with the rendered charts.
   */
  renderedCharts: RenderedChartModel[];
  /**
   * Clear all rendered charts from the mocks.
   */
  clearMocks: () => void;
};

const renderedCharts: RenderedChartModel[] = [];

const mockingConfig = {
  shouldChartBeMocked: false,
};

vi.mock('@/chart', async () => {
  const actualModule = await vi.importActual<typeof import('@/chart')>('@/chart');

  return {
    ...actualModule,
    Chart: (props: ChartProps) => {
      if (!mockingConfig.shouldChartBeMocked) {
        return actualModule.Chart(props); // Use the actual Chart component
      }
      const chartModel = new RenderedChartModel(
        props,
        `Chart_${renderedCharts.length} (${props.chartType})`,
      );
      renderedCharts.push(chartModel);
      return chartModel.render();
    },
  };
});

const mockChartComponent = () => {
  clearMocks();
  mockingConfig.shouldChartBeMocked = true;
};

const unmockChartComponent = () => {
  mockingConfig.shouldChartBeMocked = false;
};

const clearMocks = () => {
  renderedCharts.splice(0, renderedCharts.length);
};

export const chartMocksManager: ChartMocksManager = {
  mockChartComponent,
  unmockChartComponent,
  clearMocks,
  renderedCharts,
};
