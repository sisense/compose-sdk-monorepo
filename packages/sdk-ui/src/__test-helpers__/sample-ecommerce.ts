import {
  Dimension,
  DateDimension,
  Attribute,
  createAttribute,
  createDateDimension,
  createDimension,
} from '@sisense/sdk-data';

export const DataSource = 'Sample ECommerce';

interface BrandDimension extends Dimension {
  Brand: Attribute;
  BrandID: Attribute;
}
export const Brand = createDimension({
  name: 'Brand',
  Brand: createAttribute({
    name: 'Brand',
    type: 'text-attribute',
    expression: '[Brand.Brand]',
  }),
  BrandID: createAttribute({
    name: 'BrandID',
    type: 'numeric-attribute',
    expression: '[Brand.Brand ID]',
  }),
}) as BrandDimension;

interface CategoryDimension extends Dimension {
  Category: Attribute;
  CategoryID: Attribute;
}
export const Category = createDimension({
  name: 'Category',
  Category: createAttribute({
    name: 'Category',
    type: 'text-attribute',
    expression: '[Category.Category]',
  }),
  CategoryID: createAttribute({
    name: 'CategoryID',
    type: 'numeric-attribute',
    expression: '[Category.Category ID]',
  }),
}) as CategoryDimension;

interface CommerceDimension extends Dimension {
  AgeRange: Attribute;
  BrandID: Attribute;
  CategoryID: Attribute;
  Condition: Attribute;
  Cost: Attribute;
  CountryID: Attribute;
  DateMonth: Attribute;
  Gender: Attribute;
  Quantity: Attribute;
  Revenue: Attribute;
  VisitID: Attribute;
  Date: DateDimension;
}
export const Commerce = createDimension({
  name: 'Commerce',
  AgeRange: createAttribute({
    name: 'AgeRange',
    type: 'text-attribute',
    expression: '[Commerce.Age Range]',
  }),
  BrandID: createAttribute({
    name: 'BrandID',
    type: 'numeric-attribute',
    expression: '[Commerce.Brand ID]',
  }),
  CategoryID: createAttribute({
    name: 'CategoryID',
    type: 'numeric-attribute',
    expression: '[Commerce.Category ID]',
  }),
  Condition: createAttribute({
    name: 'Condition',
    type: 'text-attribute',
    expression: '[Commerce.Condition]',
  }),
  Cost: createAttribute({
    name: 'Cost',
    type: 'numeric-attribute',
    expression: '[Commerce.Cost]',
  }),
  CountryID: createAttribute({
    name: 'CountryID',
    type: 'numeric-attribute',
    expression: '[Commerce.Country ID]',
  }),
  DateMonth: createAttribute({
    name: 'DateMonth',
    type: 'numeric-attribute',
    expression: '[Commerce.Date (Month)]',
  }),
  Gender: createAttribute({
    name: 'Gender',
    type: 'text-attribute',
    expression: '[Commerce.Gender]',
  }),
  Quantity: createAttribute({
    name: 'Quantity',
    type: 'numeric-attribute',
    expression: '[Commerce.Quantity]',
  }),
  Revenue: createAttribute({
    name: 'Revenue',
    type: 'numeric-attribute',
    expression: '[Commerce.Revenue]',
  }),
  VisitID: createAttribute({
    name: 'VisitID',
    type: 'numeric-attribute',
    expression: '[Commerce.Visit ID]',
  }),
  Date: createDateDimension({
    name: 'Date',
    expression: '[Commerce.Date (Calendar)]',
  }),
}) as CommerceDimension;

interface CountryDimension extends Dimension {
  Country: Attribute;
  CountryID: Attribute;
}
export const Country = createDimension({
  name: 'Country',
  Country: createAttribute({
    name: 'Country',
    type: 'text-attribute',
    expression: '[Country.Country]',
  }),
  CountryID: createAttribute({
    name: 'CountryID',
    type: 'numeric-attribute',
    expression: '[Country.Country ID]',
  }),
}) as CountryDimension;
