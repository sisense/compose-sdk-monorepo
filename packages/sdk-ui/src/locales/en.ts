export const translation = {
  errors: {
    componentRenderError: 'Unable to render the component',

    sisenseContextNoAuthentication: 'Authentication method is not specified',

    chartNoSisenseContext:
      'Sisense Context for Chart not found. To fix, add a dataSet to the Chart or wrap the component inside a Sisense context provider.',

    dashboardWidgetNoSisenseContext:
      'Sisense Context for Dashboard Widget not found. To fix, wrap the component inside a Sisense context provider.',

    dashboardWidgetInvalidIdentifiers:
      'Failed to retrieve widget. ' +
      'Please make sure the dashboard widget exists and is accessible.',

    executeQueryNoSisenseContext:
      'Sisense Context for query execution not found. To fix, wrap the component inside a Sisense context provider.',

    executeQueryNoDataSource: 'No dataSource provided to execute query',

    dataOptions: {
      emptyValueArray: 'Invalid dataOptions â€“ Array "value" is empty',
    },
  },
  common: {
    chartNoData: 'No Results',
  },
};
