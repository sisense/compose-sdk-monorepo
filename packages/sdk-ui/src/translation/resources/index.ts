import { translation as enDictionary, TranslationDictionary } from './en.js';
import { translation as ukDictionary } from './uk.js';

export type { TranslationDictionary };
export const PACKAGE_NAMESPACE = 'sdkUi' as const;
export const resources = {
  en: enDictionary,
  uk: ukDictionary,
};
