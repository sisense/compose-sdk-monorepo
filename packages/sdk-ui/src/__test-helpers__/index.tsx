/**
 * Common test utilities
 */
import { PropsWithChildren, ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import type { Cell, Data } from '@ethings-os/sdk-data';
import isObject from 'lodash-es/isObject';
import userEvent from '@testing-library/user-event';
import { SisenseContext, SisenseContextPayload } from '../sisense-context/sisense-context';
import { Authenticator, HttpClient } from '@ethings-os/sdk-rest-client';
import { ClientApplication } from '../app/client-application';
import { DeepPartial } from 'ts-essentials';
import { useTranslation } from 'react-i18next';
import { translation } from '../translation/resources/en';
import { CustomWidgetsProvider } from '../custom-widgets-provider';
import { MenuProvider } from '@/common/components/menu/menu-provider';

type UserSetupFn = (typeof userEvent)['setup'];
interface SetupResult extends RenderResult {
  user: ReturnType<UserSetupFn>;
}
/**
 * This helper method combines the userEvent.setup() and render() calls at the
 * start of a unit test using @testing-library/react.
 *
 * Based on recommendation by: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
 *
 * @param jsx - React component instance to test, passed directly to render()
 * @param useFakeTimers - true if Vitest fake timers are enabled
 * @returns
 */
export function setup(jsx: ReactElement, useFakeTimers = false): SetupResult {
  return {
    // Need to set "advanceTimers" if we're using Vitest's fake timers.
    // The solution in https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841
    // is for Jest, but this seems to work for Vitest as well.
    user: userEvent.setup(
      useFakeTimers ? { advanceTimers: (ms) => vi.advanceTimersByTime(ms) } : undefined,
    ),
    ...render(jsx),
  };
}

export const MockedHighchartsReact = ({
  options,
}: {
  /**
   * Highcharts chart configuration object.
   * Please refer to the Highcharts (API documentation)[https://api.highcharts.com/highcharts/].
   */
  options: object;
}) => {
  return <div>{JSON.stringify(options)}</div>;
};

type MockedSisenseContextProviderProps = PropsWithChildren<
  DeepPartial<Omit<SisenseContextPayload, 'app'>>
>;

const mockHttpClient = new HttpClient(
  'http://mock-url/sometenant?someparam=true',
  {} as Authenticator,
  'mock-env',
);
export const MockedSisenseContextProvider = ({
  children,
  tracking = {
    enabled: true,
  },
  isInitialized = true,
  errorBoundary,
}: MockedSisenseContextProviderProps) => {
  return (
    <SisenseContext.Provider
      value={{
        app: {
          httpClient: mockHttpClient,
        } as ClientApplication,
        tracking: {
          enabled: tracking.enabled ?? true,
          packageName: tracking.packageName ?? 'sdk-ui',
        },
        isInitialized,
        errorBoundary: { showErrorBox: true, ...errorBoundary },
      }}
    >
      <MenuProvider>
        <CustomWidgetsProvider>{children}</CustomWidgetsProvider>
      </MenuProvider>
    </SisenseContext.Provider>
  );
};

/**
 * Extends the provided data by blurring specified rows.
 *
 * @param {Data} data - The original data object.
 * @param {number[]} blurRowsIndexes - An array of row indexes to be blurred.
 * @returns {Data} The data object with blurred rows.
 */
export const withBlurredRows = (data: Data, blurRowsIndexes: number[]) => ({
  ...data,
  rows: data.rows.map((row, rowIndex) =>
    row.map((value) => {
      const cell: Cell = isObject(value) ? value : { data: value };

      if (blurRowsIndexes.includes(rowIndex)) {
        return { ...cell, blur: true };
      }
      return cell;
    }),
  ),
});

/**
 * Mocking the i18next translation functionality for tests.
 */
export const setupI18nMock = () => {
  vi.mock('react-i18next', async () => {
    const original: { useTranslation: typeof useTranslation } = await vi.importActual(
      'react-i18next',
    );

    // Helper function to get nested translations safely
    const getNestedTranslation = (key: string): string | undefined => {
      return key.split('.').reduce((accumulator: any, currentKey: string) => {
        if (accumulator && typeof accumulator === 'object' && currentKey in accumulator) {
          return accumulator[currentKey];
        }
        return undefined;
      }, translation) as string | undefined;
    };

    return {
      ...original,
      useTranslation: () => ({
        ...original.useTranslation(),
        t: vi.fn((key, options?: Record<string, any>) => {
          const translatedValue = getNestedTranslation(key);
          if (translatedValue && options) {
            // Replace variables in the translated string
            return translatedValue.replace(/{{(.*?)}}/g, (_, varName) => {
              const trimmedVarName = varName.trim();
              return options[trimmedVarName] || `{{${trimmedVarName}}}`; // Fallback if variable is not found
            });
          }
          return translatedValue ?? key; // Fallback to the key if translation is not found
        }),
      }),
    };
  });
};
