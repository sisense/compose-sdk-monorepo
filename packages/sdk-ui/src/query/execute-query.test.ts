/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { type ClientApplication } from '..';
import {
  executeQuery,
  executeCsvQuery,
  createExecuteQueryCacheKey,
  executePivotQuery,
} from './execute-query';
import { translation } from '@/translation/resources/en';
import * as DM from '@/__test-helpers__/sample-ecommerce';
import { filterFactory, measureFactory } from '@sisense/sdk-data';

const app = {
  queryClient: {
    executeQuery: vi.fn(),
    executeCsvQuery: vi.fn(),
    executePivotQuery: vi.fn(),
  },
};

describe('executeQuery', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });
  describe(`executeQuery`, () => {
    it('should throw "seconds level supported for live datasource only" error', async () => {
      app.queryClient.executeQuery.mockReturnValue({
        resultPromise: Promise.reject({ message: 'SecondsLevelIsNotSupportedException' }),
      });

      await expect(async () => {
        await executeQuery(
          {
            dataSource: DM.DataSource,
            dimensions: [DM.Commerce.Date.Seconds],
            measures: [],
            filters: [filterFactory.members(DM.Commerce.Gender, ['Female'])],
            highlights: [filterFactory.members(DM.Commerce.AgeRange, ['18-24'])],
          },
          app as unknown as ClientApplication,
        );
      }).rejects.toThrow(translation.errors.secondsDateTimeLevelSupportedOnlyForLive);
    });
    it('should call executeQuery from queryClient', async () => {
      app.queryClient.executeQuery.mockReturnValue({
        resultPromise: Promise.resolve({}),
      });

      const result = await executeQuery(
        {
          dataSource: DM.DataSource,
          dimensions: [DM.Commerce.Condition],
          measures: [measureFactory.sum(DM.Commerce.Cost)],
          filters: [filterFactory.members(DM.Commerce.AgeRange, ['18-24'], { guid: 'test-id' })],
          highlights: [filterFactory.members(DM.Commerce.Gender, ['Male'], { guid: 'test-id' })],
        },
        app as unknown as ClientApplication,
      );

      expect(result).toEqual({});
      expect(app.queryClient.executeQuery).toHaveBeenCalledOnce();
      expect(app.queryClient.executeQuery.mock.calls[0]).toMatchSnapshot();
    });
    it('should throw error in dataSource and defaultDataSource are not provided', async () => {
      await expect(async () => {
        await executeQuery(
          {
            dataSource: undefined,
            dimensions: [],
            measures: [],
            filters: [],
            highlights: [],
          },
          app as unknown as ClientApplication,
        );
      }).rejects.toThrow('No dataSource provided to execute query');
    });
  });
  describe('executeCsvQuery', () => {
    it('should call executeCsvQuery from queryClient and return Blob', async () => {
      app.queryClient.executeCsvQuery.mockReturnValue({
        resultPromise: Promise.resolve(new Blob()),
      });

      const result = await executeCsvQuery(
        {
          dataSource: DM.DataSource,
          dimensions: [DM.Commerce.Condition],
          measures: [measureFactory.sum(DM.Commerce.Cost)],
          filters: [filterFactory.members(DM.Commerce.AgeRange, ['18-24'])],
          highlights: [filterFactory.members(DM.Commerce.Gender, ['Male'])],
        },
        app as unknown as ClientApplication,
      );

      expect(result).toBeInstanceOf(Blob);
    });
  });
  describe('executePivotQuery', () => {
    it('should call executePivotQuery from queryClient with correct args', async () => {
      app.queryClient.executePivotQuery.mockReturnValue({
        resultPromise: Promise.resolve({}),
      });

      const result = await executePivotQuery(
        {
          dataSource: DM.DataSource,
          rows: [DM.Commerce.Condition],
          columns: [DM.Commerce.AgeRange],
          values: [measureFactory.sum(DM.Commerce.Cost)],
          grandTotals: {},
          filters: [filterFactory.members(DM.Commerce.Gender, ['Female'], { guid: 'test-id' })],
          highlights: [filterFactory.members(DM.Commerce.AgeRange, ['18-24'], { guid: 'test-id' })],
        },
        app as unknown as ClientApplication,
      );

      expect(result).toEqual({});
      expect(app.queryClient.executePivotQuery).toHaveBeenCalledOnce();
      expect(app.queryClient.executePivotQuery.mock.calls[0]).toMatchSnapshot();
    });

    it('should throw error in dataSource and defaultDataSource are not provided', async () => {
      await expect(async () => {
        await executePivotQuery(
          {
            dataSource: undefined,
            rows: [],
            columns: [],
            values: [],
            grandTotals: {},
            filters: [],
            highlights: [],
          },
          app as unknown as ClientApplication,
        );
      }).rejects.toThrow('No dataSource provided to execute query');
    });
  });
  describe('createExecuteQueryCacheKey', () => {
    it('should return cache key', () => {
      const cacheKey = createExecuteQueryCacheKey(
        {
          dataSource: DM.DataSource,
          dimensions: [DM.Commerce.Condition],
          measures: [measureFactory.sum(DM.Commerce.Cost)],
          filters: [filterFactory.members(DM.Commerce.AgeRange, ['18-24'], { guid: 'test-id' })],
          highlights: [filterFactory.members(DM.Commerce.Gender, ['Male'], { guid: 'test-id' })],
        },
        app as unknown as ClientApplication,
      );

      expect(cacheKey).toMatchSnapshot();
    });
  });
});
