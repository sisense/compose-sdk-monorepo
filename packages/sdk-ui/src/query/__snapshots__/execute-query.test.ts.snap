// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`executeQuery > createExecuteQueryCacheKey > should return cache key 1`] = `"{"metadata":[{"jaql":{"title":"Condition","dim":"[Commerce.Condition]","datatype":"text"}},{"jaql":{"title":"sum Cost","dim":"[Commerce.Cost]","datatype":"numeric","agg":"sum"}},{"jaql":{"title":"Age Range","dim":"[Commerce.Age Range]","datatype":"text","filter":{"members":["18-24"]},"datasource":{"title":"Sample ECommerce","live":false}},"panel":"scope"},{"jaql":{"title":"Gender","dim":"[Commerce.Gender]","datatype":"text","filter":{"members":["Male"]},"datasource":{"title":"Sample ECommerce","live":false}},"panel":"scope"}],"datasource":{"title":"Sample ECommerce","live":false},"by":"ComposeSDK","queryGuid":""}"`;

exports[`executeQuery > executePivotQuery > should call executePivotQuery from queryClient with correct args 1`] = `
[
  {
    "columnsAttributes": [
      {
        "__serializable": "DimensionalElement",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "desc": "",
        "expression": "[Commerce.Age Range]",
        "name": "Age Range",
        "type": "text-attribute",
      },
    ],
    "count": undefined,
    "dataSource": "Sample ECommerce",
    "filterRelations": undefined,
    "filters": [
      {
        "__serializable": "DimensionalElement",
        "attribute": {
          "__serializable": "DimensionalElement",
          "dataSource": {
            "live": false,
            "title": "Sample ECommerce",
          },
          "desc": "",
          "expression": "[Commerce.Gender]",
          "name": "Gender",
          "type": "text-attribute",
        },
        "dataSource": undefined,
        "desc": "",
        "filterType": "members",
        "members": [
          "Female",
        ],
        "name": "6765f4132d6",
        "type": "filter",
      },
    ],
    "grandTotals": {},
    "highlights": [
      {
        "__serializable": "DimensionalElement",
        "attribute": {
          "__serializable": "DimensionalElement",
          "dataSource": {
            "live": false,
            "title": "Sample ECommerce",
          },
          "desc": "",
          "expression": "[Commerce.Age Range]",
          "name": "Age Range",
          "type": "text-attribute",
        },
        "dataSource": undefined,
        "desc": "",
        "filterType": "members",
        "members": [
          "18-24",
        ],
        "name": "84fa8d5c2b2",
        "type": "filter",
      },
    ],
    "measures": [
      {
        "__serializable": "DimensionalElement",
        "aggregation": "sum",
        "attribute": {
          "__serializable": "DimensionalElement",
          "dataSource": {
            "live": false,
            "title": "Sample ECommerce",
          },
          "desc": "",
          "expression": "[Commerce.Cost]",
          "name": "Cost",
          "type": "numeric-attribute",
        },
        "dataSource": undefined,
        "desc": "",
        "name": "sum Cost",
        "sort": 0,
        "type": "basemeasure",
      },
    ],
    "offset": undefined,
    "rowsAttributes": [
      {
        "__serializable": "DimensionalElement",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "desc": "",
        "expression": "[Commerce.Condition]",
        "name": "Condition",
        "type": "text-attribute",
      },
    ],
  },
  undefined,
]
`;

exports[`executeQuery > executeQuery > should call executeQuery from queryClient 1`] = `
[
  {
    "attributes": [
      {
        "__serializable": "DimensionalElement",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "desc": "",
        "expression": "[Commerce.Condition]",
        "name": "Condition",
        "type": "text-attribute",
      },
    ],
    "count": undefined,
    "dataSource": "Sample ECommerce",
    "filterRelations": undefined,
    "filters": [
      {
        "__serializable": "DimensionalElement",
        "attribute": {
          "__serializable": "DimensionalElement",
          "dataSource": {
            "live": false,
            "title": "Sample ECommerce",
          },
          "desc": "",
          "expression": "[Commerce.Age Range]",
          "name": "Age Range",
          "type": "text-attribute",
        },
        "dataSource": undefined,
        "desc": "",
        "filterType": "members",
        "members": [
          "18-24",
        ],
        "name": "84fa8d5c2b2",
        "type": "filter",
      },
    ],
    "highlights": [
      {
        "__serializable": "DimensionalElement",
        "attribute": {
          "__serializable": "DimensionalElement",
          "dataSource": {
            "live": false,
            "title": "Sample ECommerce",
          },
          "desc": "",
          "expression": "[Commerce.Gender]",
          "name": "Gender",
          "type": "text-attribute",
        },
        "dataSource": undefined,
        "desc": "",
        "filterType": "members",
        "members": [
          "Male",
        ],
        "name": "9d2957eadd5",
        "type": "filter",
      },
    ],
    "measures": [
      {
        "__serializable": "DimensionalElement",
        "aggregation": "sum",
        "attribute": {
          "__serializable": "DimensionalElement",
          "dataSource": {
            "live": false,
            "title": "Sample ECommerce",
          },
          "desc": "",
          "expression": "[Commerce.Cost]",
          "name": "Cost",
          "type": "numeric-attribute",
        },
        "dataSource": undefined,
        "desc": "",
        "name": "sum Cost",
        "sort": 0,
        "type": "basemeasure",
      },
    ],
    "offset": undefined,
    "ungroup": undefined,
  },
  undefined,
]
`;
